File: apps/desktop/src/health.ts
import { invoke } from '@tauri-apps/api/core';

export const HEALTH_COMMAND = 'health';

export type LifecycleState = 'ready' | 'not_initialized' | 'disabled';

export interface ComponentHealth {
  state: LifecycleState;
  reason: string;
}

export interface HealthReport {
  api_version: 'deskgraph.health.v1';
  product: 'DeskGraph';
  app_version: string;
  status: 'ok';
  platform: {
    os: string;
    architecture: string;
  };
  database: ComponentHealth;
  providers: {
    ocr: ComponentHealth;
    embeddings: ComponentHealth;
    local_llm: ComponentHealth;
  };
  privacy: {
    local_only_default: boolean;
    network_required: boolean;
    filesystem_locations_included: boolean;
    authorized_scope_count: number;
  };
}

type InvokeCommand = (command: string) => Promise<unknown>;

const lifecycleStates = new Set<LifecycleState>(['ready', 'not_initialized', 'disabled']);

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function isComponentHealth(value: unknown): value is ComponentHealth {
  return (
    isRecord(value) &&
    typeof value.state === 'string' &&
    lifecycleStates.has(value.state as LifecycleState) &&
    typeof value.reason === 'string'
  );
}

export function parseHealthReport(value: unknown): HealthReport {
  if (!isRecord(value) || !isRecord(value.platform) || !isRecord(value.providers)) {
    throw new Error('Invalid health response');
  }

  if (!isRecord(value.privacy)) {
    throw new Error('Invalid health response');
  }

  const valid =
    value.api_version === 'deskgraph.health.v1' &&
    value.product === 'DeskGraph' &&
    typeof value.app_version === 'string' &&
    value.status === 'ok' &&
    typeof value.platform.os === 'string' &&
    typeof value.platform.architecture === 'string' &&
    isComponentHealth(value.database) &&
    isComponentHealth(value.providers.ocr) &&
    isComponentHealth(value.providers.embeddings) &&
    isComponentHealth(value.providers.local_llm) &&
    typeof value.privacy.local_only_default === 'boolean' &&
    typeof value.privacy.network_required === 'boolean' &&
    typeof value.privacy.filesystem_locations_included === 'boolean' &&
    typeof value.privacy.authorized_scope_count === 'number';

  if (!valid) {
    throw new Error('Invalid health response');
  }

  return value as unknown as HealthReport;
}

export async function loadHealthReport(
  invokeCommand: InvokeCommand = (command) => invoke(command),
): Promise<HealthReport> {
  const response = await invokeCommand(HEALTH_COMMAND);
  return parseHealthReport(response);
}

export function lifecycleLabel(state: LifecycleState): string {
  switch (state) {
    case 'ready':
      return 'Ready';
    case 'not_initialized':
      return 'Not initialized';
    case 'disabled':
      return 'Disabled';
  }
}

File: crates/domain/src/health.rs
use serde::Serialize;

pub const HEALTH_API_VERSION: &str = "deskgraph.health.v1";

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum LifecycleState {
    Ready,
    NotInitialized,
    Disabled,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct ComponentHealth {
    pub state: LifecycleState,
    pub reason: &'static str,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct PlatformHealth {
    pub os: &'static str,
    pub architecture: &'static str,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct ProviderHealth {
    pub ocr: ComponentHealth,
    pub embeddings: ComponentHealth,
    pub local_llm: ComponentHealth,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct PrivacyHealth {
    pub local_only_default: bool,
    pub network_required: bool,
    pub filesystem_locations_included: bool,
    pub authorized_scope_count: u32,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct HealthReport {
    pub api_version: &'static str,
    pub product: &'static str,
    pub app_version: &'static str,
    pub status: &'static str,
    pub platform: PlatformHealth,
    pub database: ComponentHealth,
    pub providers: ProviderHealth,
    pub privacy: PrivacyHealth,
}

#[must_use]
pub fn collect_health() -> HealthReport {
    let provider_disabled = || ComponentHealth {
        state: LifecycleState::Disabled,
        reason: "optional_provider_not_configured",
    };

    HealthReport {
        api_version: HEALTH_API_VERSION,
        product: "DeskGraph",
        app_version: env!("CARGO_PKG_VERSION"),
        status: "ok",
        platform: PlatformHealth {
            os: std::env::consts::OS,
            architecture: std::env::consts::ARCH,
        },
        database: ComponentHealth {
            state: LifecycleState::NotInitialized,
            reason: "manifest_database_pending_m1",
        },
        providers: ProviderHealth {
            ocr: provider_disabled(),
            embeddings: provider_disabled(),
            local_llm: provider_disabled(),
        },
        privacy: PrivacyHealth {
            local_only_default: true,
            network_required: false,
            filesystem_locations_included: false,
            authorized_scope_count: 0,
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn health_is_honest_before_manifest_initialization() {
        let report = collect_health();

        assert_eq!(report.status, "ok");
        assert_eq!(report.database.state, LifecycleState::NotInitialized);
        assert_eq!(report.providers.ocr.state, LifecycleState::Disabled);
        assert_eq!(report.privacy.authorized_scope_count, 0);
        assert!(!report.privacy.network_required);
    }

    #[test]
    fn serialized_health_uses_the_stable_privacy_safe_schema() {
        let value = serde_json::to_value(collect_health()).expect("health must serialize");

        assert_eq!(value["api_version"], HEALTH_API_VERSION);
        assert_eq!(value["database"]["state"], "not_initialized");
        assert_eq!(value["providers"]["local_llm"]["state"], "disabled");
        assert_eq!(value["privacy"]["filesystem_locations_included"], false);
    }

    #[test]
    fn serialized_health_does_not_leak_filesystem_locations() {
        let serialized = serde_json::to_string(&collect_health()).expect("health must serialize");

        assert!(!serialized.contains("/Users/"));
        assert!(!serialized.contains("C:\\Users\\"));
        assert!(!serialized.contains("HOME"));
    }
}
